filelog receiver panics when delete_after_read is true

### Component(s)

receiver/filelog

### What happened?

## Description

When setting "delete_after_read: true" within the `filelogreceiver`, the collector panics with

    panic: runtime error: invalid memory address or nil pointer dereference
    [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1013b6521]


## Steps to Reproduce


1. Create a folder `data`
2. Start the opentelemetry contrib collector: `./otelcol --config config.yaml --feature-gates filelog.allowFileDeletion`
3. Collector logs `fileconsumer/file.go:132	finding files: no files match the configured criteria`
4. Create file `data/file.ndjson` with following contents:

```json
{"time":1707821387.123,"event":"This is a test: 7","host":"mbp","source":"/Users/user/somefile.log","sourcetype":"someval","index":"idx2","fields":{"orig_sourcetype":"test_messages"}}
{"time":1707821387.123,"event":"This is a test:: 8","host":"mbp","source":"/Users/user/somefile.log","sourcetype":"someval","index":"idx1"}
``` 

5. Collector logs `fileconsumer/file.go:261	Started watching file...`
6. Collector deletes the file `data/file.ndjson` as expected
7. Panic

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1013c07e1]

goroutine 23 [running]:
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/reader.(*Reader).ReadToEnd(0xc00041a070, {0x1017a6cb0, 0xc000518af0})
	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.94.0/fileconsumer/internal/reader/reader.go:49 +0x61
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer.(*Manager).preConsume.func1(0xc00041a070)
	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.94.0/fileconsumer/file_other.go:47 +0xc7
created by github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer.(*Manager).preConsume in goroutine 12
	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.94.0/fileconsumer/file_other.go:45 +0x393
```

## Expected Result
The monitored data file is deleted, execution continues.

## Actual Result
The monitored data file is deleted, a panic due to a nil pointer dereference.

This happens within [Reader.ReadToEnd()](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/449b6a41196bb1702b4d55c7420debaecb2a82e0/pkg/stanza/fileconsumer/internal/reader/reader.go#L48) as the `Reader.file` pointer is nil due to the operations of `Reader.delete()` calling `Reader.Close()`




### Collector version

v0.94.0

### Environment information

## Environment
MacOS Sonoma 14.2.1 
go version go1.21.7 darwin/amd64


### OpenTelemetry Collector configuration

```yaml
receivers:
  filelog:
    include: [ 'data/*.ndjson' ] # Adjust path as needed
    include_file_name: false
    include_file_path: true
    max_log_size: 10Mib
    encoding: utf-8
    delete_after_read: true
    start_at: beginning
    operators:
      - type: json_parser
        parse_from: body
        timestamp:
          parse_from: attributes.time
          layout_type: epoch
          layout: s.us

exporters:
  debug:
    verbosity: detailed
    sampling_initial: 5
    sampling_thereafter: 200

service:
  pipelines:
    logs:
      receivers: [filelog]
      exporters: [debug]
```


### Log output

```shell
2024-02-22T15:04:10.667+0100	info	service@v0.94.0/telemetry.go:59	Setting up own telemetry...
2024-02-22T15:04:10.667+0100	info	service@v0.94.0/telemetry.go:104	Serving metrics	{"address": ":8888", "level": "Basic"}
2024-02-22T15:04:10.667+0100	info	exporter@v0.94.1/exporter.go:275	Development component. May change in the future.	{"kind": "exporter", "data_type": "logs", "name": "debug"}
2024-02-22T15:04:10.668+0100	info	service@v0.94.0/service.go:140	Starting otelcol-custom...	{"Version": "1.0.0", "NumCPU": 8}
2024-02-22T15:04:10.668+0100	info	extensions/extensions.go:34	Starting extensions...
2024-02-22T15:04:10.668+0100	info	adapter/receiver.go:45	Starting stanza receiver	{"kind": "receiver", "name": "filelog", "data_type": "logs"}
2024-02-22T15:04:10.669+0100	warn	fileconsumer/file.go:46	finding files: no files match the configured criteria	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-02-22T15:04:10.669+0100	info	service@v0.94.0/service.go:166	Everything is ready. Begin running and processing data.
2024-02-22T15:04:10.669+0100	warn	localhostgate/featuregate.go:63	The default endpoints for all servers in components will change to use localhost instead of 0.0.0.0 in a future version. Use the feature gate to preview the new default.	{"feature gate ID": "component.UseLocalHostAsDefaultHost"}
2024-02-22T15:04:10.869+0100	warn	fileconsumer/file.go:132	finding files: no files match the configured criteria	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-02-22T15:04:11.070+0100	warn	fileconsumer/file.go:132	finding files: no files match the configured criteria	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-02-22T15:04:11.270+0100	warn	fileconsumer/file.go:132	finding files: no files match the configured criteria	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-02-22T15:04:11.470+0100	warn	fileconsumer/file.go:132	finding files: no files match the configured criteria	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-02-22T15:04:11.670+0100	warn	fileconsumer/file.go:132	finding files: no files match the configured criteria	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-02-22T15:04:11.870+0100	warn	fileconsumer/file.go:132	finding files: no files match the configured criteria	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-02-22T15:04:12.070+0100	warn	fileconsumer/file.go:132	finding files: no files match the configured criteria	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-02-22T15:04:12.269+0100	warn	fileconsumer/file.go:132	finding files: no files match the configured criteria	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-02-22T15:04:12.469+0100	warn	fileconsumer/file.go:132	finding files: no files match the configured criteria	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer"}
2024-02-22T15:04:19.070+0100	info	fileconsumer/file.go:261	Started watching file	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer", "path": "data/file.ndjson"}
2024-02-22T15:04:19.170+0100	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 2}
2024-02-22T15:04:19.170+0100	info	ResourceLog #0
Resource SchemaURL:
ScopeLogs #0
ScopeLogs SchemaURL:
InstrumentationScope
LogRecord #0
ObservedTimestamp: 2024-02-22 14:04:19.071309 +0000 UTC
Timestamp: 2024-02-13 10:49:47.123 +0000 UTC
SeverityText:
SeverityNumber: Unspecified(0)
Body: Str({"time":1707821387.123,"event":"THAT IS PAOLOS ia-lx-test DATA NUMBER: 7 - The data string must be larger so that a data set generates more bytes","host":"mbp-de-cric","source":"/Users/crichter/Applications/splunk/var/spool/splunk/9750b17f362275ad_4f7a27434c189849_events.stash_new","sourcetype":"ia-lx-test_test_messages","index":"ia-lx-test-ingest-actions-test-p","fields":{"orig_sourcetype":"test_messages"}})
Attributes:
     -> fields: Map({"orig_sourcetype":"test_messages"})
     -> time: Double(1707821387.123)
     -> log.file.path: Str(data/file.ndjson)
     -> event: Str(THAT IS PAOLOS ia-lx-test DATA NUMBER: 7 - The data string must be larger so that a data set generates more bytes)
     -> host: Str(mbp-de-cric)
     -> source: Str(/Users/crichter/Applications/splunk/var/spool/splunk/9750b17f362275ad_4f7a27434c189849_events.stash_new)
     -> sourcetype: Str(ia-lx-test_test_messages)
     -> index: Str(ia-lx-test-ingest-actions-test-p)
Trace ID:
Span ID:
Flags: 0
LogRecord #1
ObservedTimestamp: 2024-02-22 14:04:19.071815 +0000 UTC
Timestamp: 2024-02-13 10:49:47.123 +0000 UTC
SeverityText:
SeverityNumber: Unspecified(0)
Body: Str({"time":1707821387.123,"event":"THAT IS PAOLOS lx9114 DATA NUMBER: 8 - The data string must be larger so that a data set generates more bytes","host":"mbp-de-cric","source":"/Users/crichter/Applications/splunk/var/spool/splunk/9750b17f362275ad_4f7a27434c189849_events.stash_new","sourcetype":"lx9114_test_messages","index":"lx9114"})
Attributes:
     -> log.file.path: Str(data/file.ndjson)
     -> source: Str(/Users/crichter/Applications/splunk/var/spool/splunk/9750b17f362275ad_4f7a27434c189849_events.stash_new)
     -> sourcetype: Str(lx9114_test_messages)
     -> index: Str(lx9114)
     -> time: Double(1707821387.123)
     -> event: Str(THAT IS PAOLOS lx9114 DATA NUMBER: 8 - The data string must be larger so that a data set generates more bytes)
     -> host: Str(mbp-de-cric)
Trace ID:
Span ID:
Flags: 0
	{"kind": "exporter", "data_type": "logs", "name": "debug"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1013c07e1]

goroutine 23 [running]:
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/reader.(*Reader).ReadToEnd(0xc00041a070, {0x1017a6cb0, 0xc000518af0})
	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.94.0/fileconsumer/internal/reader/reader.go:49 +0x61
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer.(*Manager).preConsume.func1(0xc00041a070)
	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.94.0/fileconsumer/file_other.go:47 +0xc7
created by github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer.(*Manager).preConsume in goroutine 12
	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.94.0/fileconsumer/file_other.go:45 +0x393
```


### Additional context

I was able to fix the issue by patching Reader.ReadToEnd(ctx context.Context) with 

```go
// ReadToEnd will read until the end of the file
func (r *Reader) ReadToEnd(ctx context.Context) {
    if r.file == nil {
        return
    }

```
However, I am not well versed in the otel source code and am not sure this is the correct thing to do.

--- a/.chloggen/pkg-stanza-fileconsumer-darfix.yaml
+++ b/.chloggen/pkg-stanza-fileconsumer-darfix.yaml
@@ -1,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: filelogreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Fix bug where delete_after_read would cause panic
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [31383]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: []

--- a/pkg/stanza/fileconsumer/file.go
+++ b/pkg/stanza/fileconsumer/file.go
@@ -67,7 +67,6 @@ func (m *Manager) Start(persister operator.Persister) error {
 
 func (m *Manager) closePreviousFiles() {
 	// m.previousPollFiles -> m.knownFiles[0]
-
 	for r, _ := m.previousPollFiles.Pop(); r != nil; r, _ = m.previousPollFiles.Pop() {
 		m.knownFiles[0].Add(r.Close())
 	}
@@ -171,7 +170,7 @@ func (m *Manager) consume(ctx context.Context, paths []string) {
 	m.Debug("Consuming files", zap.Strings("paths", paths))
 	m.makeReaders(paths)
 
-	m.preConsume(ctx)
+	m.readLostFiles(ctx)
 
 	// read new readers to end
 	var wg sync.WaitGroup

--- a/pkg/stanza/fileconsumer/file_other.go
+++ b/pkg/stanza/fileconsumer/file_other.go
@@ -15,7 +15,12 @@ import (
 // Take care of files which disappeared from the pattern since the last poll cycle
 // this can mean either files which were removed, or rotated into a name not matching the pattern
 // we do this before reading existing files to ensure we emit older log lines before newer ones
-func (m *Manager) preConsume(ctx context.Context) {
+func (m *Manager) readLostFiles(ctx context.Context) {
+	if m.readerFactory.DeleteAtEOF {
+		// Lost files are not expected when delete_at_eof is enabled
+		// since we are deleting the files before they can become lost.
+		return
+	}
 	lostReaders := make([]*reader.Reader, 0, m.previousPollFiles.Len())
 OUTER:
 	for _, oldReader := range m.previousPollFiles.Get() {

--- a/pkg/stanza/fileconsumer/file_test.go
+++ b/pkg/stanza/fileconsumer/file_test.go
@@ -985,16 +985,14 @@ func TestDeleteAfterRead(t *testing.T) {
 	linesPerFile := 10
 	totalLines := files * linesPerFile
 
-	expectedTokens := make([][]byte, 0, totalLines)
-	actualTokens := make([][]byte, 0, totalLines)
-
 	tempDir := t.TempDir()
 	temps := make([]*os.File, 0, files)
 	for i := 0; i < files; i++ {
 		temps = append(temps, filetest.OpenTemp(t, tempDir))
 	}
 
-	// Write logs to each file
+	expectedTokens := make([][]byte, 0, totalLines)
+	actualTokens := make([][]byte, 0, totalLines)
 	for i, temp := range temps {
 		for j := 0; j < linesPerFile; j++ {
 			line := filetest.TokenWithLength(100)
@@ -1014,6 +1012,35 @@ func TestDeleteAfterRead(t *testing.T) {
 	sink := emittest.NewSink(emittest.WithCallBuffer(totalLines))
 	operator := testManagerWithSink(t, cfg, sink)
 	operator.persister = testutil.NewUnscopedMockPersister()
+	operator.poll(context.Background())
+	actualTokens = append(actualTokens, sink.NextTokens(t, totalLines)...)
+
+	require.ElementsMatch(t, expectedTokens, actualTokens)
+
+	for _, temp := range temps {
+		_, err := os.Stat(temp.Name())
+		require.True(t, os.IsNotExist(err))
+	}
+
+	// Make more files to ensure deleted files do not cause problems on next poll
+	temps = make([]*os.File, 0, files)
+	for i := 0; i < files; i++ {
+		temps = append(temps, filetest.OpenTemp(t, tempDir))
+	}
+
+	expectedTokens = make([][]byte, 0, totalLines)
+	actualTokens = make([][]byte, 0, totalLines)
+	for i, temp := range temps {
+		for j := 0; j < linesPerFile; j++ {
+			line := filetest.TokenWithLength(200)
+			message := fmt.Sprintf("%s %d %d", line, i, j)
+			_, err := temp.WriteString(message + "\n")
+			require.NoError(t, err)
+			expectedTokens = append(expectedTokens, []byte(message))
+		}
+		require.NoError(t, temp.Close())
+	}
+
 	operator.poll(context.Background())
 	actualTokens = append(actualTokens, sink.NextTokens(t, totalLines)...)
 

--- a/pkg/stanza/fileconsumer/file_windows.go
+++ b/pkg/stanza/fileconsumer/file_windows.go
@@ -9,7 +9,8 @@ import (
 	"context"
 )
 
-func (m *Manager) preConsume(ctx context.Context) {
+// Noop on windows because we close files immediately after reading.
+func (m *Manager) readLostFiles(ctx context.Context) {
 }
 
 // On windows, we close files immediately after reading because they cannot be moved while open.

--- a/pkg/stanza/fileconsumer/internal/reader/reader.go
+++ b/pkg/stanza/fileconsumer/internal/reader/reader.go
@@ -114,7 +114,7 @@ func (r *Reader) ReadToEnd(ctx context.Context) {
 
 // Delete will close and delete the file
 func (r *Reader) delete() {
-	r.Close()
+	r.close()
 	if err := os.Remove(r.fileName); err != nil {
 		r.logger.Errorf("could not delete %s", r.fileName)
 	}
@@ -122,6 +122,13 @@ func (r *Reader) delete() {
 
 // Close will close the file and return the metadata
 func (r *Reader) Close() *Metadata {
+	r.close()
+	m := r.Metadata
+	r.Metadata = nil
+	return m
+}
+
+func (r *Reader) close() {
 	if r.file != nil {
 		if err := r.file.Close(); err != nil {
 			r.logger.Debugw("Problem closing reader", zap.Error(err))
@@ -134,9 +141,6 @@ func (r *Reader) Close() *Metadata {
 			r.logger.Errorw("Failed to stop header pipeline", zap.Error(err))
 		}
 	}
-	m := r.Metadata
-	r.Metadata = nil
-	return m
 }
 
 // Read from the file and update the fingerprint if necessary

